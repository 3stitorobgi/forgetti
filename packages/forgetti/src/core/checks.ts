import * as t from '@babel/types';
import type * as babel from '@babel/core';
import type { ComponentNode, StateContext } from './types';

export function getImportSpecifierName(specifier: t.ImportSpecifier): string {
  if (specifier.imported.type === 'Identifier') {
    return specifier.imported.name;
  }
  return specifier.imported.value;
}

export function isComponent(node: t.Node): node is ComponentNode {
  switch (node.type) {
    case 'ArrowFunctionExpression':
    case 'FunctionExpression':
    case 'FunctionDeclaration':
      return true;
    default:
      return false;
  }
}

export function isComponentNameValid(
  ctx: StateContext,
  node: ComponentNode,
  checkName: boolean,
): boolean {
  if (checkName) {
    if (node.type !== 'ArrowFunctionExpression') {
      return (
        !!node.id
        && (
          ctx.filters.component.test(node.id.name)
          || (!!ctx.filters.hook && ctx.filters.hook.test(node.id.name))
        )
      );
    }
    return false;
  }
  return true;
}

type TypeFilter<V extends t.Node> = (node: t.Node) => node is V;

export function isPathValid<V extends t.Node>(
  path: unknown,
  key: TypeFilter<V>,
): path is babel.NodePath<V> {
  return key((path as babel.NodePath).node);
}

export type NestedExpression =
  | t.ParenthesizedExpression
  | t.TypeCastExpression
  | t.TSAsExpression
  | t.TSSatisfiesExpression
  | t.TSNonNullExpression
  | t.TSInstantiationExpression
  | t.TSTypeAssertion;

export function isNestedExpression(node: t.Node): node is NestedExpression {
  switch (node.type) {
    case 'ParenthesizedExpression':
    case 'TypeCastExpression':
    case 'TSAsExpression':
    case 'TSSatisfiesExpression':
    case 'TSNonNullExpression':
    case 'TSTypeAssertion':
    case 'TSInstantiationExpression':
      return true;
    default:
      return false;
  }
}

/** Check whether a Node is CallExpression and a React Hook call */
export function isExpressionAndHook(
  ctx: StateContext,
  node: t.Expression | t.SpreadElement,
): boolean {
  switch (node.type) {
    case 'CallExpression':
      // A simple check to see if "node.callee" is an Identifier
      // TypeScript can infer the type of "node.callee" to Identifier or V8IntrinsicIdentifier
      if (!('name' in node.callee)) {
        return false;
      }

      // Check if callee is a react hook
      if (!ctx.filters.hook) {
        return false;
      }
      return ctx.filters.hook.test(node.callee.name);
    case 'SpreadElement':
      return isExpressionAndHook(ctx, node.argument);
    case 'ArrayExpression':
      return node.elements.some((element) => {
        if (!element) return false;
        return isExpressionAndHook(ctx, element);
      });
    case 'ObjectExpression':
      return node.properties.some((property) => {
        if (!property) return false;
        if (property.type === 'ObjectProperty') {
          return (
            (t.isExpression(property.value) && isExpressionAndHook(ctx, property.value))
            || (t.isSpreadElement(property.key) && isExpressionAndHook(ctx, property.key))
          );
        }
        if (property.type === 'SpreadElement') {
          return isExpressionAndHook(ctx, property);
        }
        return false;
      });
    case 'TemplateLiteral':
      return node.expressions.some((expression) => {
        if (!expression) return false;
        if (!t.isExpression(expression)) return false;
        return isExpressionAndHook(ctx, expression);
      });
    case 'BinaryExpression':
      return (
        (t.isExpression(node.left) && isExpressionAndHook(ctx, node.left))
      || (t.isExpression(node.right) && isExpressionAndHook(ctx, node.right))
      );
    case 'AssignmentExpression':
      return (t.isExpression(node.right) && isExpressionAndHook(ctx, node.right));
    default:
      return false;
  }
}
