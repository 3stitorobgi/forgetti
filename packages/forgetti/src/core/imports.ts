import type { ImportRegistration } from './presets';

export const RUNTIME_EQUALS: ImportRegistration = {
  name: '$$equals',
  source: 'forgetti/runtime',
  kind: 'named',
};

export const RUNTIME_CACHE: ImportRegistration = {
  name: '$$cache',
  source: 'forgetti/runtime',
  kind: 'named',
};
