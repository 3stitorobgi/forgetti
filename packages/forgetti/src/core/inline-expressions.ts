/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import type * as babel from '@babel/core';
import type { ComponentNode } from './types';

function isInValidExpression(path: babel.NodePath): boolean {
  let current = path.parentPath;
  let prev = path;
  while (current) {
    if (
      current.isConditionalExpression()
      && (
        current.get('consequent').node === prev.node
        || current.get('alternate').node === prev.node
      )
    ) {
      return true;
    }
    if (
      current.isLogicalExpression()
      && current.get('right').node === prev.node
    ) {
      return true;
    }
    prev = current;
    current = current.parentPath;
  }
  return false;
}

export function inlineExpressions(
  path: babel.NodePath<ComponentNode>,
): void {
  path.traverse({
    Expression(p) {
      if (p.getFunctionParent() === path && p.isIdentifier()) {
        const binding = p.scope.getBinding(p.node.name);
        if (binding && binding.referenced && binding.referencePaths.length === 1) {
          switch (binding.kind) {
            case 'const':
            case 'let':
            case 'var': {
              // move the node to the reference
              const ref = binding.referencePaths[0];
              if (
                binding.path.isVariableDeclarator()
                && binding.path.get('id').isIdentifier()
                && binding.path.scope.getBlockParent() === ref.scope.getBlockParent()
                && binding.path.node.init
                && !isInValidExpression(ref)
              ) {
                ref.replaceWith(binding.path.node.init);
                binding.path.remove();
              }
            }
              break;
            default:
              break;
          }
        }
      }
    },
  });
  path.scope.crawl();
}
