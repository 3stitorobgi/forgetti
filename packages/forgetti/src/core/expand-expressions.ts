/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import * as t from '@babel/types';
import type * as babel from '@babel/core';
import type { ComponentNode, StateContext } from './types';
import { isPathValid } from './checks';
import unwrapNode from './unwrap-node';

export function expandExpressions(
  ctx: StateContext,
  path: babel.NodePath<ComponentNode>,
): void {
  if (path.node.type === 'ArrowFunctionExpression') {
    if (path.node.body.type !== 'BlockStatement') {
      path.node.body = t.blockStatement(
        [t.returnStatement(path.node.body)],
      );
    }
  }

  const hoistedVars = new Set();

  path.traverse({
    AssignmentExpression(p) {
      const parent = p.getFunctionParent();

      if (parent === path) {
        if (
          p.parentPath.isVariableDeclarator()
          && p.parentPath.node.id.type === 'Identifier'
          && hoistedVars.has(p.parentPath.node.id.name)
        ) {
          return;
        }
        const statement = p.getStatementParent();
        if (statement) {
          const id = p.scope.generateUidIdentifier('hoisted');
          hoistedVars.add(id.name);
          statement.insertBefore(
            t.variableDeclaration(
              'let',
              [t.variableDeclarator(id, p.node)],
            ),
          );
          p.replaceWith(id);
        }
      }
    },
    CallExpression(p) {
      const parent = p.getFunctionParent();

      if (parent === path) {
        if (
          p.parentPath.isVariableDeclarator()
          && p.parentPath.node.id.type === 'Identifier'
          && hoistedVars.has(p.parentPath.node.id.name)
        ) {
          return;
        }
        const calleePath = p.get('callee');
        if (isPathValid(calleePath, t.isExpression)) {
          let isHook = false;
          const trueID = unwrapNode(calleePath.node, t.isIdentifier);
          if (trueID) {
            const binding = p.scope.getBindingIdentifier(trueID.name);
            if (binding) {
              const registration = ctx.registrations.hooks.get(binding);
              if (registration) {
                isHook = true;
              }
            }
            if (ctx.filters.hook?.test(trueID.name)) {
              isHook = true;
            }
            // Check if callee is potentially a namespace import
          }
          const trueMember = unwrapNode(calleePath.node, t.isMemberExpression);
          if (
            trueMember
            && !trueMember.computed
            && t.isIdentifier(trueMember.property)
          ) {
            const obj = unwrapNode(trueMember.object, t.isIdentifier);
            if (obj) {
              const binding = path.scope.getBindingIdentifier(obj.name);
              if (binding) {
                const registrations = ctx.registrations.hooksNamespaces.get(binding);
                if (registrations) {
                  let registration: typeof registrations[0];
                  for (let i = 0, len = registrations.length; i < len; i += 1) {
                    registration = registrations[i];
                    if (registration.name === trueMember.property.name) {
                      isHook = true;
                    }
                  }
                }
              }
            }
            if (ctx.filters.hook?.test(trueMember.property.name)) {
              isHook = true;
            }
          }

          if (isHook) {
            const statement = p.getStatementParent();
            if (statement) {
              const id = p.scope.generateUidIdentifier('hoisted');
              hoistedVars.add(id.name);
              statement.insertBefore(
                t.variableDeclaration(
                  'let',
                  [t.variableDeclarator(id, p.node)],
                ),
              );
              p.replaceWith(id);
            }
          }
        }
      }
    },
  });

  path.scope.crawl();
}
