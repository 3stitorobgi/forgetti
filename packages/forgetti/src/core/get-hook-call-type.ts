/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import type * as babel from '@babel/core';
import * as t from '@babel/types';
import { isPathValid } from './checks';
import type { HookIdentity } from './presets';
import unwrapNode from './unwrap-node';
import type { StateContext } from './types';

type HookCallType =
  | HookIdentity
  | 'custom'
  | 'none';

export function getHookCallType(
  ctx: StateContext,
  path: babel.NodePath<t.CallExpression | t.OptionalCallExpression>,
): HookCallType {
  const callee = path.get('callee');
  if (isPathValid(callee, t.isV8IntrinsicIdentifier)) {
    return 'none';
  }
  const trueID = unwrapNode(callee.node, t.isIdentifier);
  if (trueID) {
    const binding = path.scope.getBindingIdentifier(trueID.name);
    if (binding) {
      const registration = ctx.registrations.hooks.get(binding);
      if (registration) {
        return registration.type;
      }
    }
    return ctx.filters.hook && ctx.filters.hook.test(trueID.name) ? 'custom' : 'none';
  }
  // Check if callee is potentially a namespace import
  const trueMember = unwrapNode(callee.node, t.isMemberExpression);
  if (
    trueMember
      && !trueMember.computed
      && t.isIdentifier(trueMember.property)
  ) {
    const obj = unwrapNode(trueMember.object, t.isIdentifier);
    if (obj) {
      const binding = path.scope.getBindingIdentifier(obj.name);
      if (binding) {
        const registrations = ctx.registrations.hooksNamespaces.get(binding);
        if (registrations) {
          let registration: typeof registrations[0];
          for (let i = 0, len = registrations.length; i < len; i += 1) {
            registration = registrations[i];
            if (registration.name === trueMember.property.name) {
              return registration.type;
            }
          }
        }
      }
    }
    return ctx.filters.hook && ctx.filters.hook.test(trueMember.property.name)
      ? 'custom'
      : 'none';
  }
  return 'none';
}
