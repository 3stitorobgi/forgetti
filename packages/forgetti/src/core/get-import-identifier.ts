import * as babel from '@babel/core';
import { addDefault, addNamed } from '@babel/helper-module-imports';
import { ImportRegistration } from './presets';
import { StateContext } from './types';

export default function getImportIdentifier(
  ctx: StateContext,
  path: babel.NodePath,
  definition: ImportRegistration,
) {
  const target = `${definition.source}[${definition.name}]`;
  const current = ctx.hooks.get(target);
  if (current) {
    return current;
  }
  const newID = (definition.kind === 'named')
    ? addNamed(path, definition.name, definition.source)
    : addDefault(path, definition.source);
  ctx.hooks.set(target, newID);
  return newID;
}
