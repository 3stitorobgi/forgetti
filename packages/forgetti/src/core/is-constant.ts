/* eslint-disable @typescript-eslint/no-unnecessary-condition */
/* eslint-disable @typescript-eslint/no-use-before-define */
import type * as babel from '@babel/core';
import * as t from '@babel/types';
import { isNestedExpression, isPathValid } from './checks';
import type OptimizerScope from './optimizer-scope';
import getForeignBindings, { isForeignBinding } from './get-foreign-bindings';
import type { ComponentNode, StateContext } from './types';
import unwrapNode from './unwrap-node';

interface OptimizerInstance {
  ctx: StateContext;
  scope: OptimizerScope;
  path: babel.NodePath<ComponentNode>;
}

function isTemplateLiteralConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.TemplateLiteral>,
): boolean {
  const expressions = path.get('expressions');
  for (let i = 0, len = expressions.length; i < len; i++) {
    const expr = expressions[i];
    if (isPathValid(expr, t.isExpression)) {
      if (!isConstant(instance, expr)) {
        return false;
      }
    } else {
      return false;
    }
  }
  return true;
}

function isIdentifierConstant(
  instance: OptimizerInstance,
  path: babel.NodePath,
  node: t.Identifier,
): boolean {
  switch (node.name) {
    case 'undefined':
    case 'NaN':
    case 'Infinity':
      return true;
    default: {
      if (isForeignBinding(instance.path, path, node.name)) {
        return true;
      }
      const binding = path.scope.getBindingIdentifier(node.name);
      if (binding) {
        return instance.scope.hasConstant(binding);
      }
      return true;
    }
  }
}

function isMemberExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.MemberExpression | t.OptionalMemberExpression>,
): boolean {
  if (isConstant(instance, path.get('object'))) {
    if (path.node.computed) {
      const property = path.get('property');
      if (isPathValid(property, t.isExpression)) {
        return isConstant(instance, property);
      }
    }
    return true;
  }
  return false;
}

function isConditionalExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.ConditionalExpression>,
): boolean {
  return isConstant(instance, path.get('test'))
    && isConstant(instance, path.get('consequent'))
    && isConstant(instance, path.get('alternate'));
}

function isBinaryExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.BinaryExpression>,
): boolean {
  const left = path.get('left');
  const right = path.get('right');
  return isPathValid(left, t.isExpression)
    && isConstant(instance, left)
    && isConstant(instance, right);
}

function isLogicalExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.LogicalExpression>,
): boolean {
  return isConstant(instance, path.get('left'))
    && isConstant(instance, path.get('right'));
}

function isUnaryExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.UnaryExpression>,
): boolean {
  switch (path.node.operator) {
    case 'throw':
    case 'delete':
      return false;
    default:
      return isConstant(instance, path.get('argument'));
  }
}

function isCallExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.CallExpression | t.OptionalCallExpression>,
): boolean {
  const callee = path.get('callee');
  if (isPathValid(callee, t.isExpression)) {
    const trueID = unwrapNode(callee.node, t.isIdentifier);
    if (trueID) {
      const binding = path.scope.getBindingIdentifier(trueID.name);
      if (binding) {
        const registration = instance.ctx.registrations.hooks.get(binding);
        if (registration) {
          return false;
        }
        if (instance.ctx.filters.hook?.test(trueID.name)) {
          return false;
        }
      }
    }
    const trueMember = unwrapNode(callee.node, t.isMemberExpression);
    if (
      trueMember
      && !trueMember.computed
      && t.isIdentifier(trueMember.property)
    ) {
      const obj = unwrapNode(trueMember.object, t.isIdentifier);
      if (obj) {
        const binding = path.scope.getBindingIdentifier(obj.name);
        if (binding) {
          const registrations = instance.ctx.registrations.hooksNamespaces.get(binding);
          if (registrations) {
            let registration: typeof registrations[0];
            for (let i = 0, len = registrations.length; i < len; i += 1) {
              registration = registrations[i];
              if (registration.name === trueMember.property.name) {
                return false;
              }
            }
          }
        }
      }
      if (instance.ctx.filters.hook?.test(trueMember.property.name)) {
        return false;
      }
    }
    if (!isConstant(instance, callee)) {
      return false;
    }
  }
  const args = path.get('arguments');
  for (let i = 0, len = args.length; i < len; i++) {
    const arg = args[i];
    if (isPathValid(arg, t.isExpression) && !isConstant(instance, arg)) {
      return false;
    }
    if (isPathValid(arg, t.isSpreadElement) && !isConstant(instance, arg.get('argument'))) {
      return false;
    }
  }
  return true;
}

function isFunctionExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.FunctionExpression | t.ArrowFunctionExpression>,
): boolean {
  const bindings = getForeignBindings(path);
  for (let i = 0, len = bindings.length; i < len; i++) {
    const binding = bindings[i];
    if (!isIdentifierConstant(instance, path, binding)) {
      return false;
    }
  }
  return true;
}

function isArrayExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.ArrayExpression | t.TupleExpression>,
): boolean {
  const elements = path.get('elements');
  for (let i = 0, len = elements.length; i < len; i++) {
    const element = elements[i];
    if (isPathValid(element, t.isSpreadElement)) {
      if (!isConstant(instance, element.get('argument'))) {
        return false;
      }
    }
    if (isPathValid(element, t.isExpression)) {
      if (!isConstant(instance, element)) {
        return false;
      }
    }
  }
  return true;
}

function isObjectExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.ObjectExpression | t.RecordExpression>,
): boolean {
  const properties = path.get('properties');
  for (let i = 0, len = properties.length; i < len; i++) {
    const property = properties[i];
    if (isPathValid(property, t.isSpreadElement)) {
      if (!isConstant(instance, property.get('argument'))) {
        return false;
      }
    }
    if (isPathValid(property, t.isObjectProperty)) {
      if (property.node.computed) {
        const key = property.get('key');
        if (isPathValid(key, t.isExpression) && !isConstant(instance, key)) {
          return false;
        }
      }
      const value = property.get('value');
      if (isPathValid(value, t.isExpression) && !isConstant(instance, value)) {
        return false;
      }
    }
    if (isPathValid(property, t.isObjectMethod)) {
      // TODO
    }
  }
  return true;
}

function isNewExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.NewExpression>,
): boolean {
  const callee = path.get('callee');
  if (isPathValid(callee, t.isExpression) && isConstant(instance, callee)) {
    const args = path.get('arguments');
    for (let i = 0, len = args.length; i < len; i++) {
      const arg = args[i];
      if (isPathValid(arg, t.isExpression) && !isConstant(instance, arg)) {
        return false;
      }
      if (isPathValid(arg, t.isSpreadElement) && !isConstant(instance, arg.get('argument'))) {
        return false;
      }
    }
    return true;
  }
  return false;
}

function isSequenceExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.SequenceExpression>,
): boolean {
  const exprs = path.get('expressions');
  for (let i = 0, len = exprs.length; i < len; i++) {
    if (!isConstant(instance, exprs[i])) {
      return false;
    }
  }
  return true;
}

function isTaggedTemplateExpressionConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.TaggedTemplateExpression>,
): boolean {
  return isConstant(instance, path.get('tag'))
    && isConstant(instance, path.get('quasi'));
}

function isJSXChildrenConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.JSXFragment | t.JSXElement>,
): boolean {
  const children = path.get('children');
  for (let i = 0, len = children.length; i < len; i++) {
    const child = children[i];
    if (isPathValid(child, t.isJSXElement) && !isJSXElementConstant(instance, child)) {
      return false;
    }
    if (isPathValid(child, t.isJSXFragment) && !isJSXChildrenConstant(instance, child)) {
      return false;
    }
    if (isPathValid(child, t.isJSXSpreadChild) && !isConstant(instance, child.get('expression'))) {
      return false;
    }
    if (isPathValid(child, t.isJSXExpressionContainer)) {
      const expr = child.get('expression');
      if (isPathValid(expr, t.isExpression) && !isConstant(instance, expr)) {
        return false;
      }
    }
  }
  return true;
}

function isJSXNameConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.JSXIdentifier | t.JSXMemberExpression | t.JSXNamespacedName>,
): boolean {
  if (isPathValid(path, t.isJSXNamespacedName)) {
    return true;
  }
  if (isPathValid(path, t.isJSXMemberExpression)) {
    return isJSXNameConstant(instance, path.get('object'));
  }
  if (isPathValid(path, t.isJSXIdentifier)) {
    if (/^[A-Z]/.test(path.node.name)) {
      return isForeignBinding(instance.path, path, path.node.name);
    }
  }
  return true;
}

function isJSXOpeningElementConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.JSXOpeningElement>,
): boolean {
  if (isJSXNameConstant(instance, path.get('name'))) {
    const attrs = path.get('attributes');
    for (let i = 0, len = attrs.length; i < len; i++) {
      const attr = attrs[i];
      if (isPathValid(attr, t.isJSXAttribute)) {
        const value = attr.get('value');
        if (isPathValid(value, t.isJSXElement) && !isJSXElementConstant(instance, value)) {
          return false;
        }
        if (isPathValid(value, t.isJSXFragment) && !isJSXChildrenConstant(instance, value)) {
          return false;
        }
        if (isPathValid(value, t.isJSXExpressionContainer)) {
          const expr = value.get('expression');
          if (isPathValid(expr, t.isExpression) && !isConstant(instance, expr)) {
            return false;
          }
        }
      }
      if (isPathValid(attr, t.isJSXSpreadAttribute) && !isConstant(instance, attr.get('argument'))) {
        return false;
      }
    }
    return true;
  }
  return false;
}

function isJSXElementConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.JSXElement>,
): boolean {
  return isJSXOpeningElementConstant(instance, path.get('openingElement'))
    && isJSXChildrenConstant(instance, path);
}

export default function isConstant(
  instance: OptimizerInstance,
  path: babel.NodePath<t.Expression>,
): boolean {
  if (isPathValid(path, isNestedExpression)) {
    return isConstant(instance, path.get('expression'));
  }
  if (isPathValid(path, t.isLiteral)) {
    if (isPathValid(path, t.isTemplateLiteral)) {
      return isTemplateLiteralConstant(instance, path);
    }
    return true;
  }
  if (isPathValid(path, t.isIdentifier)) {
    return isIdentifierConstant(instance, path, path.node);
  }
  if (isPathValid(path, t.isMemberExpression) || isPathValid(path, t.isOptionalMemberExpression)) {
    return isMemberExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isConditionalExpression)) {
    return isConditionalExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isBinaryExpression)) {
    return isBinaryExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isLogicalExpression)) {
    return isLogicalExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isUnaryExpression)) {
    return isUnaryExpressionConstant(instance, path);
  }
  if (
    isPathValid(path, t.isCallExpression)
    || isPathValid(path, t.isOptionalCallExpression)
  ) {
    return isCallExpressionConstant(instance, path);
  }
  if (
    isPathValid(path, t.isFunctionExpression)
    || isPathValid(path, t.isArrowFunctionExpression)
  ) {
    return isFunctionExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isAssignmentExpression)) {
    // TODO
    return false;
  }
  if (isPathValid(path, t.isArrayExpression) || isPathValid(path, t.isTupleExpression)) {
    return isArrayExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isObjectExpression) || isPathValid(path, t.isRecordExpression)) {
    return isObjectExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isNewExpression)) {
    return isNewExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isSequenceExpression)) {
    return isSequenceExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isTaggedTemplateExpression)) {
    return isTaggedTemplateExpressionConstant(instance, path);
  }
  if (isPathValid(path, t.isJSXFragment)) {
    return isJSXChildrenConstant(instance, path);
  }
  if (isPathValid(path, t.isJSXElement)) {
    return isJSXElementConstant(instance, path);
  }
  if (isPathValid(path, t.isImport)) {
    return true;
  }
  if (isPathValid(path, t.isMetaProperty)) {
    return true;
  }
  return false;
}
