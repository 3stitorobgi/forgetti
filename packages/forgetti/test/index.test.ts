/* eslint-disable no-template-curly-in-string */
import * as babel from '@babel/core';
import { describe, expect, it } from 'vitest';
import plugin, { Options } from '../src';

const options: Options = {
  shouldCheckComponentName: true,
  memo: {
    name: 'useMemo',
    source: 'react',
    kind: 'named',
  },
  hooks: [
    {
      type: 'memo',
      name: 'useMemo',
      source: 'react',
      kind: 'named',
    },
    {
      type: 'callback',
      name: 'useCallback',
      source: 'react',
      kind: 'named',
    },
    {
      type: 'effect',
      name: 'useEffect',
      source: 'react',
      kind: 'named',
    },
    {
      type: 'effect',
      name: 'useLayoutEffect',
      source: 'react',
      kind: 'named',
    },
  ],
  hocs: [],
};

async function compile(code: string) {
  const result = await babel.transformAsync(code, {
    plugins: [
      [plugin, options],
    ],
    parserOpts: {
      plugins: [
        'jsx',
      ],
    },
  });

  return result?.code ?? '';
}

describe('example', () => {
  it('should run', async () => {
    const code = `
import { useMemo, useEffect } from 'react';
function Example(props) {
  const message = props.greeting + ', ' + props.receiver;
  const example = useMemo(() => props.format(message), [message]);

  useEffect(() => {
    console.log(example);
  }, [example]);
}
    
`;
    expect(await compile(code)).toMatchSnapshot();
  });
});
