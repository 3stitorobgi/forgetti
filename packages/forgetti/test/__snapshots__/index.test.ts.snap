// Vitest Snapshot v1

exports[`statements > should optimize for-of statements 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
function Example(props) {
  let _c = _useMemo(() => new Array(5), []),
    _eq = Object.is(_c[0], props),
    _v4 = _eq ? _c[0] : _c[0] = props,
    _v5 = _eq ? _c[1] : _c[1] = _v4.a,
    _v3;
  if (_v5) {
    let _c2 = _c[2] ||= new Array(1),
      _v6 = _eq ? _c2[0] : _c2[0] = _v4.b;
    _v3 = _v6;
  } else _v3 = _v5;
  let _v2;
  if (_v3) {
    let _c3 = _c[3] ||= new Array(1),
      _v7 = _eq ? _c3[0] : _c3[0] = _v4.c;
    _v2 = _v7;
  } else _v2 = _v3;
  let _v;
  if (_v2) {
    let _c4 = _c[4] ||= new Array(1),
      _v8 = _eq ? _c4[0] : _c4[0] = _v4.d;
    _v = _v8;
  } else _v = _v2;
  return _v;
}"
`;
